### ISSUE TITLE
GPS precision field [almost ready to close]

### ISSUE THREAD

**Issue by sacrevert** [BiologicalRecordsCentre/BSBI-Card-and-PlantPortal-DEPRECATED-#19]
Allow standard field alongside finer-scale grid ref to capture GPS precision (this should be mapable as for records coming through the iRecord app).

Nested in #3

**Comment by johnvanbreda** [comment]
Will this be an extra column in the species entry grid? If so, what title do you want, and should it be shown in the standard grid columns, or as an extra field under the + button?

**Comment by sacrevert** [comment]
Yes, as it is only relevant for edited records with a 'Finer grid ref.' Title: GPS precision. It would probably be best to have on the standard grid (immediately right of Finer grid ref), and have 'Determiner' moved to under the + button, as this will only rarely be used.

**Comment by DavidRoy** [comment]
I agree with both of Oli's suggestions: 
add GPS precision field to after 'Finer grid ref.' field
move 'Determiner' field between 'Altitude' and 'Comments' fields

**Comment by sacrevert** [comment]
hi @johnvanbreda do you have time to do this one next? I think it is on the list of ones that we considered of 'high' importance. In addition to the addition of the field, and the rearrangement of the existing fields described above. I wonder whether this issue also has other implications. For example, records coming through the iRecord app are mapped with the GPS uncertainty -- presumably precision captured in the way described above should be stored and used in the same way as this information coming through the app (but distinguished from 'preset' grid-based precisions (e.g. a monad obviously has a default precision of 1km). @japonicus may wish to comment

**Comment by japonicus** [comment]
related to https://github.com/Indicia-Team/warehouse/issues/253

**Comment by johnvanbreda** [comment]
Code for this is now done in the client helpers develop branch so will be deployed at the next feature release. Will need to set the @spatialRefPrecisionAttrId parameter on the grid configuration when the code is upgraded.

**Comment by sacrevert** [comment]
@johnvanbreda I realise this is only waiting on the code to be upgraded, but I'm just wondering whether the move of the determiner field is included in this code?

>Yes, as it is only relevant for edited records with a 'Finer grid ref.' Title: GPS precision. It would probably be best to have on the standard grid (immediately right of Finer grid ref), and have 'Determiner' moved to under the + button, as this will only rarely be used. [[above](https://github.com/BiologicalRecordsCentre/BSBI-Card-and-PlantPortal/issues/19#issuecomment-262762991)]

**Comment by johnvanbreda** [comment]
Added GPS precision and moved determiner to + button.

**Comment by japonicus** [comment]
Sorry have only just noticed the implementation of this.

'GPS Precision' obviously only makes sense if the user is recording a centroid point derived from a GPS device.  The prominence now given to this field is confusing if instead (as will commonly be the case) the precise grid-ref denotes a grid square.

It seems unclear whether, having entered a 6-fig grid square I'm also supposed to fill in the precision field. (I would know not to, but a real user of the recording form might be mislead.)

It's not clear whether the precision field refers to diameter or radius.

There are no units specified ('metres' should be explicitly stated) and it shouldn't be possible to enter anything other than blank or a positive non-zero numbers in the field (non-numeric input is silently converted to '0' and zeros are permitted). If I enter e.g. "1km" then the value is truncated to 1 (referring to 1m).

If a precision is entered but there is no precise grid-ref then the precision is ignored - this should probably trigger an error.

It's possible to enter a precision value that is more precise than the specified grid-reference square.  This is problematic as it's not clear what the precision means in that context. This should not be permitted.

For records without precision, the field is defaulting to '0' in the iRecord summary view of the record (and also if the record is edited again).  The default should be blank, because zero is not a valid precision.

I think there is need to have a way to indicate whether a centroid or grid-square is intended (as, at present, grid-square may be the more common use-case). The precision field should only be displayed for centroid references.

**Comment by sacrevert** [comment]
I agree with all of Tom's points, and think that all the checks that he recommends should be implemented. There is clearly potential for a lot of confusion here, I doubt all that many recorders will understand the distinction between centroid and grid square precisions (however it is important and should be kept).

**Question:** Is it possible/simple to have a '?' icon next to the column header (or title if it is hidden under '+') which results in a pop-up explanation?

**Comment by johnvanbreda** [comment]
> 'GPS Precision' obviously only makes sense if the user is recording a centroid point derived from a GPS device. The prominence now given to this field is confusing if instead (as will commonly be the case) the precise grid-ref denotes a grid square.

Although I agree, the implementation was such that you could only enter a gps coordinate if the main sample was also set to GPS lat long, which would always result in the warning about the finer spatial ref being outside the parent sample's one. I've changed the code now so that you can enter OSGB or OSIE grid reference for the parent sample and a GPS lat long or a grid square for the finer grid ref. GPS precision will disable itself if the finer grid ref given refers to a grid square.

> It's not clear whether the precision field refers to diameter or radius.

I've added a hint to the control to make it clear it is a radius, though in my experience don't most GPS or mobile devices always returns the radius?

> There are no units specified ('metres' should be explicitly stated) and it shouldn't be possible to enter anything other than blank or a positive non-zero numbers in the field (non-numeric input is silently converted to '0' and zeros are permitted). If I enter e.g. "1km" then the value is truncated to 1 (referring to 1m).

I've added the unit to the column title and made the input control reject any characters you type in that are not numeric. I don't see why a zero should not be accepted - this would infer a point where you are confident of the exact location.

> It's possible to enter a precision value that is more precise than the specified grid-reference square. This is problematic as it's not clear what the precision means in that context. This should not be permitted.

This is no longer relevant as you can't enter a precision for a grid square.

> For records without precision, the field is defaulting to '0' in the iRecord summary view of the record (and also if the record is edited again). The default should be blank, because zero is not a valid precision.

Having made the other changes listed here, I cannot reproduce this.

Still to do:

> If a precision is entered but there is no precise grid-ref then the precision is ignored - this should probably trigger an error.

> Is it possible/simple to have a '?' icon next to the column header (or title if it is hidden under '+') which results in a pop-up explanation?

Though, maybe the hint on the control is enough?

**Comment by japonicus** [comment]
John, thanks very much for these changes. My concerns were that people often don't grasp of how precision works and are inclined to try to fill in every blank field on a form.  It's possible that I was overstating the issues, but it would have been problematic if people misunderstood and misused the precision field.

I want recording to reach the point where GPS use is ubiquitous and were precision is applied properly. At present I often encounter quite pervasive mistrust of GPS-derived centroids by recorders, driven in part by their experience of widespread misuse of centroids in the past.

Would it be possible for the GPS precision field to be enabled if a sufficiently precise grid-reference is entered (e.g. 10-figure, and perhaps also 8-figure). Someone doing grid-square based recording will probably have their GPS set to show OS grid-refs so won't have ready access to a lat/lng. (e.g. my etrex unit shows either lat/lng or grid-ref but not both at the same time and switching between the two systems is fiddly).

**Comment by sacrevert** [comment]
I think Tom's idea about enabling the precision field if something is sub-6 figure OS is sensible. The current approach doesn't really work, because many people will have their GPS set to OS, and will want to report precisions relative to this (I would not switch my etrex unit to lat/long to record a location + precision).

**Comment by johnvanbreda** [comment]
It's getting a bit arbitrary as to when this field should be enabled IMHO and I can see some users asking why the field enables in some cases and not others. Would a more transparent solution not be to leave it enabled all the time but rename the column to "Precision if GPS used (m)"?

**Comment by sacrevert** [comment]
I think perhaps it would be less arbitrary if it was only enabled for 10-figure grid refs. These are what is likely to be noted by recorders with a GPS unit anyway (in OS grid format, with a precision). 8-figs can be assumed to represent a square, given that websites such as [this ](http://www.cucaera.co.uk/grp)provide square-refs down to that resolution.

So, if @japonicus is in agreement, I think that:

1. We restrict precision entries to 10-fig OS refs.
2. We hide the field under the 'plus' expansion (without renaming it), swapping it with the comment field, which is far more likely to be used.
3. Provide a pop-up guidance sentence for this item that says:
"This is the precision (normally given as a radius in metres) reported by most GPS devices or phones. This field is only enabled for 10-figure OS grid references. These represent 1 m accuracy, however, this is normally associated with some degree of error, which is the GPS precision recorded here. Precise grid references reported to a lower accuracy than 10-figures are assumed to represent the corresponding OS square (e.g. a 100 m or 10 m square), and so additional precision information is not required."

**Comment by japonicus** [comment]
I'd be very happy with @sacrevert's suggestions.

**Comment by johnvanbreda** [comment]
Changes now done and deployed.
